Add int24 support

sunng87 opened the pull request:
Hi Zach,

This is a patch for gloss to support int24 and uint24 as primitives. 3-byte integer are supported in Netty, and also used in [Diameter Protocol](http://en.wikipedia.org/wiki/Diameter_%28protocol%29#Packet_format). Please review this patch because it's a little complex to convert bytes<->int<->uint . The algorithm is stolen from Netty's Buffer API.

And, Merry Christmas !

KingMob commented:
@sunng87 I don't suppose you have any tests after a decade, do you? Would you be willing to write some?

sunng87 commented:
@KingMob No problem! I believe this feature is still useful after a decade.

KingMob commented:
...wow, are you still using Gloss 10 years later?

KingMob commented on src/gloss/data/primitives.clj:
```diff
@@ -61,6 +61,38 @@
   [x]
   (.longValue (bigint x)))
 
+(defn bytes-umedium
```
For consistency, these fns should have arrows in their name, a la `bytes->umedium`

KingMob commented on src/gloss/data/primitives.clj:
```diff
@@ -61,6 +61,38 @@
+         (bit-shift-right x 8)
+         (bit-shift-right x 0)])))
+
+(defn getMedium [buffer]
+  (let [bytes (byte-array 3)]
+    (.get buffer bytes)
+    (bytes-umedium bytes)))
+
+(defn putMedium [buffer value]
+  (let [bytes (medium-bytes value)]
+    (.put buffer bytes)))
```
Since these aren't actually methods on `ByteBuffer`, I'd prefer to keep them Lisp kebab-cased, like `get-medium` and `put-medium`.

sunng87 replied:
No problem. I'm also going to check if type hint is required here.

sunng87 commented:
@KingMob No... I switched to my own Netty wrapper [link](https://github.com/clojure-link/link) in ~2013 and implemented similar mechanism. But I believe the approach from gloss is real elegant. It's worthy of finishing this last few bits.

KingMob commented:
Ahh, ok. Well, since the PR was created, Gloss got better support for endian-ness; can you support that as well? 

Currently the primitive codecs look like:

```clojure
   :int32 (primitive-codec .getInt .putInt 4 identity int identity)
   :int32-le (primitive-codec .getInt .putInt 4 identity int identity :le)
   :int32-be (primitive-codec .getInt .putInt 4 identity int identity :be)
```

I hope it's not too hard to make sure the int24 version code and tests behave the same.

sunng87 commented:
Sounds good. I will take a look later today or tomorrow.

sunng87 commented:
This pull request seems to be based from a commit instead of a branch (may be early feature of github). I will close this one and create a new PR for this.
